kubeflow/notebooks/prototypes/README.md

```diff
@@ -0,0 +1,131 @@
+## Goals
+
+- provide a k8 native mechanism to spawning jupyter notebooks for users
+
+- spawn a notebook within a RBAC protected namespace for a user
+
+- set a serviceAccountName within the spawning pod similar to jupyter-notebook
+
+- use k8 native authentication for the user within the protected namespace
+
+- allow similar params as kubeflow/core/prototypes/jupyterhub.jsonnet, including PVCs
+
+
+## Design
+
+The Notebooks component is an alternative to jupyterhub. The component defines a Notebook CRD and provides a notebook-controller (based on metacontroller's [CompositeController](https://metacontroller.app/api/compositecontroller/)). The Notebook CRD schema is similar to a kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment) where it defines a [Pod Template](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template) that is used by the notebook-controller to create a Pod. An example Notebook yaml is shown below:
+
+```yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: Notebook
+metadata:
+  name: training
+  namespace: resnet50
+spec:
+  template:
```

Comment:
yes, i'll make this change. I was leaning towards Job. Thanks.

Reply:
Why Job? The semantics of Job are run until completion. I think of Jupyter more as a server; so the semantics of a statefulset or deployment seem more appropriate. I lean towards statefulset just to give the pods predictable names.

Reply:
Sure, I'll replace with a StatefulSet. 

I was thinking of ways to eliminate the long wait for notebook to come up - due to the cost of pulling the image. Are there ways to eliminate that time cost under a StatefulSet?

Reply:
nit: might want to use `deployment` since stateful sets give you some weird guarantees. Namely:

```
 Deleting and/or scaling a StatefulSet down will not delete the volumes associated with the StatefulSet. This is done to ensure data safety, which is generally more valuable than an automatic purge of all related StatefulSet resources.
```
which means if a user specifies a PVC in a notebook, creating/deleting the notebook CRD will never remove that volume...which could be argued as intended behavior? But as it stands right now, the volume management is configured outside of the CRD which I like for its simplicitly and breaking down ownership of various components.

```
StatefulSets currently require a Headless Service to be responsible for the network identity of the Pods. You are responsible for creating this Service
```
Not really an issue, more of a nit to be careful around when creating the service.

Reply:
The image pulling cost is independent of the controller type so I dont think deployment/ss matters. What we do is we run a `daemonset` with the fat notebook images as init-containers and the daemonset itself just runs the pause container. This essentially creates an image pre-fetch layer and caches it on all the kubelets. You could add a template of that here if you want to simplify people's lives.